### PR TITLE
AP_Scripting: RGB LED override binding

### PR DESCRIPTION
**Pull request by IamPete1**
This work was funded by 3DXR.

This adds a new method to AP_Notify to allow temporary overriding all other RGB led methods. The binding for this is then added to scripting. Along with a example to blink the lights. 

https://www.youtube.com/watch?v=TyiMCSSfkz0&feature=youtu.be

### CODE REVIEW

File: libraries/AP_Notify/RGBLed.h [hunk 1]
```diff
@@ -38,7 +38,12 @@ class RGBLed: public NotifyDevice {
 
     // handle LED control, only used when LED_OVERRIDE=1
     virtual void handle_led_control(const mavlink_message_t &msg) override;
-    
+
+    // RGB override
+    // override other all methods for given duration (0 = for ever)
+    // used with scripting
+    virtual void rgb_override(uint8_t r, uint8_t g, uint8_t b, uint8_t rate_hz, uint16_t duration) override;
```
Comment (tridge) [inline libraries/AP_Notify/RGBLed.h]: it would be good to give the units for duration

**Comment by WickedShell** [comment]
The question I have is do we want to just have a NTF_LED_OVERRIDE that puts all LED control in the hands of scripting? The time part of the override actually seems a bit odd to me, as it means you have to run a script on a tighter deadline to avoid reversion to old LED colors then just having full control over it. Personally I think the duration just complicates the LED side from scripts but if others want it then I can see it. (With the param stuff it gets quite easy to do temporary script LED control by just changing NTF_LED_OVERRIDE around as needed)

File: libraries/AP_Notify/RGBLed.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_Notify/RGBLed.h]: fixed, added to the header and also changed the param to duration_ms

**Comment by IamPete1** [comment]
@WickedShell I actually like the duration better than the param change, Its probability better not to encourage people to change params more than necessary using scripting. I would be tempted to move the mavlink stuff that way too, have the duration as a mavlink extension and default to some reasonable time out, 30s or some thing.  The current max duration (if you don't use 0 for for ever) is 65 seconds, not really to onerous of a call back time. May even be worth moving from ms to s for the duration, duration of less than one seconds are probably not that useful. Then we could do 18 hours (or change to a smaller variable)

**Comment by WickedShell** [comment]
I'm mostly not sold because if you miss for any reason you get a flash of a different color, if you are using the LED to indicate a safety state from your script that momentary flash could be really disturbing. I guess we could still have a NTF_LED_OVERRIDE mode that says "scripting only" and still have the duration call and when it expires just don't do anything. It just bugs me cause it means larger scripts for the general control case :) But if we want the time based one then I won't block it. Tagging it as a dev call topic incase there isn't a conclusion before then.

File: libraries/AP_Notify/RGBLed.h [hunk 2]
```diff
@@ -57,12 +60,21 @@ class RGBLed: public NotifyDevice {
     uint8_t _led_medium;
     uint8_t _led_dim;
 
-    struct {
+    struct override {
```
Comment (peterbarker) [inline libraries/AP_Notify/RGBLed.h]: Should we *really* be using a keyword here? :-)

**Comment by IamPete1** [comment]
move to a override parameter with no duration

**Comment by IamPete1** [comment]
or look at a custom version of the ObjectBuffer - ObjectBufferWithSemaphore

**Comment by IamPete1** [comment]
Moved to not having a override duration and sharing LED_OVERRIDE type 1 with mavlink, tested on CubeOrange and here2

**Comment by WickedShell** [comment]
We may want to make a variant for the future that allows `rate_hz` to be optional, but we might as well bring it in for now.

**Pull request merged**